pkg/client/typed/discovery/restmapper.go

```diff
@@ -0,0 +1,166 @@
+package discovery
+
+import (
+	"sync"
+
+	kapi "k8s.io/kubernetes/pkg/api"
+	"k8s.io/kubernetes/pkg/api/meta"
+	"k8s.io/kubernetes/pkg/api/unversioned"
+	"k8s.io/kubernetes/pkg/apimachinery/registered"
+)
+
+type discoveryRESTMapper struct {
+	discoveryClient DiscoveryInterface
+
+	delegate meta.RESTMapper
+
+	initLock sync.Mutex
+}
+
+// NewRESTMapper that initializes using the discovery APIs, relying on group ordering and preferred versions
+// to build its appropriate priorities.  Only versions are registered with API machinery are added now.
+// TODO make this work with generic resources at some point.  For now, this handles enabled and disabled resources cleanly.
+func NewRESTMapper(discoveryClient DiscoveryInterface) meta.RESTMapper {
+	return &discoveryRESTMapper{discoveryClient: discoveryClient}
+}
+
+func (d *discoveryRESTMapper) getDelegate() (meta.RESTMapper, error) {
+	d.initLock.Lock()
+	defer d.initLock.Unlock()
+
+	if d.delegate != nil {
+		return d.delegate, nil
+	}
+
+	serverGroups, err := d.discoveryClient.ServerGroups()
+	if err != nil {
+		return nil, err
+	}
+
+	// always prefer our default group for now.  The version should be discovered from discovery, but this will hold us
+	// for quite some time.
+	resourcePriority := []unversioned.GroupVersionResource{
+		{Group: kapi.GroupName, Version: meta.AnyVersion, Resource: meta.AnyResource},
+	}
+	kindPriority := []unversioned.GroupVersionKind{
+		{Group: kapi.GroupName, Version: meta.AnyVersion, Kind: meta.AnyKind},
+	}
+	groupPriority := []string{}
+
+	unionMapper := meta.MultiRESTMapper{}
+
+	for _, group := range serverGroups.Groups {
+		if len(group.Versions) == 0 {
+			continue
+		}
+		groupPriority = append(groupPriority, group.Name)
+
+		if len(group.PreferredVersion.Version) != 0 {
+			preferredVersion := unversioned.GroupVersion{Group: group.Name, Version: group.PreferredVersion.Version}
+			if registered.IsEnabledVersion(preferredVersion) {
```

Comment:
Why does it have to be registered in the client code? Should it only matter what is enabled on the server side?